### ISSUE TITLE
[MLIR][gfx906][test_conv_igemm_mlir] Failed creating module

### ISSUE THREAD

**Issue by shurale-nkn** [ROCmSoftwarePlatform/MIOpen#979]
MLIR build failed on test_conv_igemm_mlir on rocm 4.2.

Posibly because of gfx906:sramecc+:xnack-

ROCM 4.2
gfx906

http://micimaster.amd.com/blue/organizations/jenkins/MLLibs%2FMIOpen/detail/jenkins-ci-rocm-4.2/7/pipeline/129

```
11/69 Test #58: test_conv_igemm_mlir .............***Failed    8.41 sec
[  0%] Built target sqlite_memvfs
[  2%] Built target addkernels
[100%] Built target MIOpen
[100%] Built target test_conv2d
Scanning dependencies of target test_conv_igemm_mlir
MIOpen(HIP): Info [GetFindModeValueImpl] MIOPEN_FIND_MODE = NORMAL(1)
MIOpen(HIP): Info [get_device_name] Raw device name: gfx906:sramecc+:xnack-
MIOpen(HIP): Info [Handle] stream: 0x22f5bc0, device_id: 5
MIOpen(HIP): Info [BackwardDataGetWorkSpaceSize] 
MIOpen(HIP): Info [HipCompilerVersionImpl] 4.2.21155
MIOpen(HIP): Info [AmdRocmMetadataVersionDetect] ROCm MD version AMDHSA_COv3, MIOpen version 2.12.0.8dd3b8aa9
MIOpen(HIP): Info [GetEnvFindOnlySolverImpl] 98
MIOpen(HIP): Info [ForwardGetWorkSpaceSize] 
MIOpen(HIP): Info [BackwardWeightsGetWorkSpaceSize] 
/var/jenkins/workspace/LLibs_MIOpen_jenkins-ci-rocm-4.2/build/bin/test_conv2d --float --cmode conv --pmode default --group-count 1 --disable-backward-data --disable-backward-weights --input 256, 1024, 14, 14 --weights 2048, 1024, 1, 1 --pads_strides_dilations 0 0 2 2 1 1 --trans_output_pads 0 0 --in_layout NCHW --fil_layout NCHW --out_layout NCHW 
MIOpen(HIP): Info [ForwardGetWorkSpaceSize] 
MIOpen(HIP): Info [FindConvFwdAlgorithm] requestAlgoCount = 1, workspace = 0
MIOpen(HIP): Info [TryLoad] Find-db regenerating.
MIOpen(HIP): Info [FindSolutionImpl] ConvMlirIgemmFwd (not searchable)
terminate called after throwing an instance of 'miopen::Exception'
  what():  /var/jenkins/workspace/LLibs_MIOpen_jenkins-ci-rocm-4.2/src/hipoc/hipoc_program.cpp:160: Failed creating module from file /tmp/miopen-mlir_gen_igemm_conv2d_v4r4_fwd.mlir-0cb1-b1be-d889-28d0/mlir_gen_igemm_conv2d_v4r4_fwd.mlir.o hipErrorSharedObjectInitFailed
Aborted (core dumped)
test/CMakeFiles/test_conv_igemm_mlir.dir/build.make:57: recipe for target 'test/CMakeFiles/test_conv_igemm_mlir' failed
make[7]: *** [test/CMakeFiles/test_conv_igemm_mlir] Error 134
CMakeFiles/Makefile2:11508: recipe for target 'test/CMakeFiles/test_conv_igemm_mlir.dir/all' failed
make[6]: *** [test/CMakeFiles/test_conv_igemm_mlir.dir/all] Error 2
CMakeFiles/Makefile2:11515: recipe for target 'test/CMakeFiles/test_conv_igemm_mlir.dir/rule' failed
make[5]: *** [test/CMakeFiles/test_conv_igemm_mlir.dir/rule] Error 2
Makefile:1784: recipe for target 'test/CMakeFiles/test_conv_igemm_mlir.dir/rule' failed
make[4]: *** [test/CMakeFiles/test_conv_igemm_mlir.dir/rule] Error 2

```

**Comment by atamazov** [comment]
Currently the target ROCm version for MLIR feature is 4.1.1.

**Comment by jerryyin** [comment]
I synced with @atamazov briefly on this: my ask is to delay upgrading to ROCm 4.2 just as long as the June demo finishes. My estimation of the demo is that it shouldn't last longer than two weeks later on June 24th. I will leave another comment on this ticket once demo is done. Then please do what's necessary to disable the MLIR stage. (And we will enable it later when MLIR is tested and works on ROCm 4.2). Thanks!

CC: @whchung @mvermeulen

**Comment by shurale-nkn** [comment]
@jerryyin
I have another suggestion. As part of PR #970, I will install additional special rocm for MLIR.
What version of ROCm are you preparing for the June demo? 4.1.1? I can set it for MLIR, for better testing.

**Comment by atamazov** [comment]
Agreed & approved.

**Comment by shurale-nkn** [comment]
@jerryyin
Unfortunately on ROCm 4.1.1 failed too.

But i have new information.

Works fine with gpu gfx900 which is identified by the system as `gfx900:xnack-`
http://micimaster.amd.com/blue/organizations/jenkins/MLLibs%2FMIOpen/detail/jenkins-ci-rocm-4.2/14/pipeline/130

And failed on gfx906  which is identified by the system as `gfx906:sramecc+:xnack-`
http://micimaster.amd.com/blue/organizations/jenkins/MLLibs%2FMIOpen/detail/jenkins-ci-rocm-4.2/13/pipeline

And locally i have only gfx906 which is identified by the system as `gfx906:sramecc-:xnack-`
And it works.


So the sramecc status should be an issue

**Comment by atamazov** [comment]
@jerryyin 
> On MIOpen end, it will need to pass along the gcnArch to MLIR so MLIR can have the right target tripple to set in building the kernel.

[Note] Please look how TargetDescription is used to pass this info to COMGR and to offline HIP compiler.

**Comment by jerryyin** [comment]
@atamazov Thanks for the pointer. I finally have time to start working on this. I took a look at what you suggested. Do you agree with me making the following changes:

1. Store this rawName as a member of the `struct TargetPorperties{};`

https://github.com/ROCmSoftwarePlatform/MIOpen/blob/c812a33784b4081e0e733ece5632648dc4f8deb0/src/target_properties.cpp#L71-L76

2. Swap the below:

https://github.com/ROCmSoftwarePlatform/MIOpen/blob/c812a33784b4081e0e733ece5632648dc4f8deb0/src/solver/mlir_common.cpp#L68

With 

```cpp
std::string(" --arch_features ") + ctx.GetStream().GetDeviceName().GetRawName() + 
```

I think leaving MLIR to handle the rawName (and let MLIR to parse the string to get the arch + features) will be more extendible.

**Comment by atamazov** [comment]
@jerryyin Please borrow design from COMGR, like this:

```C++
#include <miopen/hip_build_utils.hpp> // for LcOptionTargetStrings
#include <miopen/rocm_features.hpp> // for ROCM_FEATURE_TARGETID_OFF
#include <miopen/target_properties.hpp> // for TargetProperties
...
static std::string GetIsaName(const miopen::TargetProperties& target)
{
#if ROCM_FEATURE_TARGETID_OFF
    return {target.Name()}; /// artem: this is what we have now, right?
#else
    const LcOptionTargetStrings lots(target);
    return {lots.targetId4Mlir};
#endif
}
```

You'll also need to add `const std::string targetId4Mlir` to `LcOptionTargetStrings` and initialize it to `(device + sramecc_reported + xnack)`.

This should keep compatibility with older ROCm versions.

**Comment by jerryyin** [comment]
@atamazov Thanks a lot! Will push for a new PR.

**Comment by jerryyin** [comment]
Closing the ticket as #1052 merged and MLIR CI bumped to ROCm 4.2. Please feel free to reopen if there's anything else needs to be done in this ticket.
